rest/src/main/java/com/redhat/ipaas/rest/v1/dao/jsondb/JsonDBDAO.java

```diff
@@ -0,0 +1,137 @@
+import com.redhat.ipaas.jsondb.JsonDB;
+import com.redhat.ipaas.rest.v1.dao.DataAccessObject;
+import com.redhat.ipaas.rest.v1.model.ListResult;
+import com.redhat.ipaas.rest.v1.model.WithId;
+import com.redhat.ipaas.rest.v1.util.Json;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+
+/**
+ * Implements a DataAccessObject using the {@see: JsonDB}.
+ */
+abstract public class JsonDBDAO<T extends WithId> implements DataAccessObject<T> {
+
+    private final JsonDB rtdb;
+
+    public JsonDBDAO(JsonDB rtdb) {
+        this.rtdb = rtdb;
+    }
+
+    abstract public String getCollectionPath();
+
+    @Override
+    public T fetch(String id) {
+        try {
+            String dbPath = getCollectionPath()+"/:"+id;
+            byte[] json = rtdb.getAsByteArray(dbPath);
+            if( json==null || json.length == 0 ) {
+                return null;
+            }
+            return Json.mapper().readValue(json, getType());
+        } catch (RuntimeException|IOException e) {
+            throw IPaasServerException.launderThrowable(e);
+        }
+    }
+
+    @Override
+    public ListResult<T> fetchAll() {
+        try {
+            // get the data out..
+            byte[] json = rtdb.getAsByteArray(getCollectionPath());
+            if( json!=null && json.length > 0 ) {
+
+                // Lets use jackson to parse the map of keys to our model instances
+                ObjectMapper mapper = Json.mapper();
+                TypeFactory typeFactory = mapper.getTypeFactory();
+                MapType mapType = typeFactory.constructMapType(LinkedHashMap.class, String.class, getType());
+                LinkedHashMap<String, T> map = mapper.readValue(json, mapType);
+
+                return ListResult.of(map.values());
+            } else {
+                return ListResult.of(Collections.EMPTY_LIST);
+            }
+        } catch (RuntimeException|IOException e) {
+            throw IPaasServerException.launderThrowable(e);
+        }
+    }
+
+    @Override
+    public T create(T entity) {
+        try {
+
+            String dbPath = getCollectionPath()+"/:"+entity.getId().get();
+
+            // Only create if it did not exist.
+            if( rtdb.exists(dbPath) ) {
+                return null;
+            }
+
+            byte[] json = Json.mapper().writeValueAsBytes(entity);
+            rtdb.set(dbPath, json);
+
+            return entity;
+
+        } catch (RuntimeException|IOException e) {
+            throw IPaasServerException.launderThrowable(e);
+        }
+    }
+
+    @Override
+    public T update(T entity) {
+        try {
+            T previousValue = this.fetch((String) entity.getId().get());
+
+            // Only update if the entity existed.
+            if( previousValue !=null ) {
+                String dbPath = getCollectionPath()+"/:"+entity.getId().get();
+                byte[] json = Json.mapper().writeValueAsBytes(entity);
```

Comment:
Same here: An update simply overrides the whole document,so the client has to provide the whole tree of objects again, but as its an update it probably has already fetched the whole tree anyway.